src/main/java/in/org/projecteka/hiu/consent/ConsentService.java

```diff
@@ -34,4 +40,32 @@ public ConsentService(ConsentManagerClient consentManagerClient,
                                 hiuProperties.getCallBackUrl()))
                                 .thenReturn(ConsentCreationResponse.builder().id(consentCreationResponse.getId()).build()));
     }
+
+    public Mono<Void> fetchConsentArtefact(String consentManagerId,
+                                           ConsentNotificationRequest consentNotificationRequest) {
+        return validateRequest(consentNotificationRequest.getConsentRequestId())
+                .flatMap(consentRequest -> validateConsentManager(consentManagerId, consentRequest)
+                        .thenMany(fetchAndStoreConsentArtefacts(consentNotificationRequest))
+                        .then());
+    }
+
+    private Flux<Void> fetchAndStoreConsentArtefacts(ConsentNotificationRequest consentNotificationRequest) {
+        return Flux.fromIterable(consentNotificationRequest.getConsents())
+                .flatMap(consentArtefactReference ->
+                        consentManagerClient.getConsentArtefact(consentArtefactReference.getId())
+                                .flatMap(consentArtefactResponse ->
+                                        consentRepository.insertConsentArtefact(consentArtefactResponse.getConsentDetail()))
+                );
+    }
+
+    private Mono<in.org.projecteka.hiu.consent.model.ConsentRequest> validateRequest(String consentRequestId) {
+        return consentRepository.get(consentRequestId)
+                .switchIfEmpty(Mono.error(consentRequestNotFound()));
+    }
+
+    private Mono<Void> validateConsentManager(String consentManagerId, in.org.projecteka.hiu.consent.model.ConsentRequest consentRequest) {
+        boolean validConsentManager = consentRequest.getPatient().getId().contains(consentManagerId);
+        if (validConsentManager) return Mono.empty();
+        return Mono.error(invalidConsentManager());
```

Comment:
if it's not making a IO call and make sense to not use Mono/Flux, always prefer to do that.